src/core/server/services/users/index.ts

```diff
@@ -360,23 +361,88 @@ export async function deactivateToken(
 }
 
 /**
- * updateUsername will update a given User's username.
+ * updateUserUsername will update the current users.
  *
  * @param mongo mongo database to interact with
+ * @param mailer mailer queue instance
  * @param tenant Tenant where the User will be interacted with
  * @param userID the User's ID that we are updating
  * @param username the username that we are setting on the User
  */
-export async function updateUsername(
+export async function updateOwnUsername(
   mongo: Db,
+  mailer: MailerQueue,
   tenant: Tenant,
-  userID: string,
+  user: User,
   username: string
 ) {
   // Validate the username.
   validateUsername(username);
 
-  return updateUserUsername(mongo, tenant.id, userID, username);
+  const lastUsernameEditAllowed = new Date();
+  const dateDiff = lastUsernameEditAllowed.getDate() - 14;
+  lastUsernameEditAllowed.setDate(dateDiff);
+
+  if (user.status.username.history.length > 1) {
+    const recentUpdate = user.status.username
+      ? user.status.username.history.find(history => {
+          return history.createdAt > lastUsernameEditAllowed;
+        })
+      : null;
+
+    if (recentUpdate) {
+      throw new UsernameUpdatedWithinWindowError(recentUpdate.createdAt);
+    }
+  }
+
+  const updated = await updateUsername(
+    mongo,
+    tenant.id,
+    user.id,
+    username,
+    user.id
+  );
+
+  if (user.email) {
+    await mailer.add({
+      tenantID: tenant.id,
+      message: {
+        to: user.email,
+      },
+      template: {
+        name: "update-username",
+        context: {
+          username: user.username!,
+          organizationName: tenant.organization.name,
+          organizationURL: tenant.organization.url,
+          organizationContactEmail: tenant.organization.contactEmail,
+        },
+      },
+    });
+  }
```

Comment:
Should add a `else` that logs that the email wasn't sent because the user didn't have an email address.